File: src/main/java/io/bloco/faker/Faker.java
package io.bloco.faker;

import java.io.IOException;
import java.io.InputStream;
import org.yaml.snakeyaml.Yaml;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.net.URL;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.bloco.faker.components.Address;
import io.bloco.faker.components.App;
import io.bloco.faker.components.Avatar;
import io.bloco.faker.components.Book;
import io.bloco.faker.components.Bool;
import io.bloco.faker.components.Business;
import io.bloco.faker.components.Color;
import io.bloco.faker.components.Commerce;
import io.bloco.faker.components.Company;
import io.bloco.faker.components.Date;
import io.bloco.faker.components.Food;
import io.bloco.faker.components.Internet;
import io.bloco.faker.components.Lorem;
import io.bloco.faker.components.Name;
import io.bloco.faker.components.Number;
import io.bloco.faker.components.PhoneNumber;
import io.bloco.faker.components.Placeholdit;
import io.bloco.faker.components.SlackEmoji;
import io.bloco.faker.components.Team;
import io.bloco.faker.components.Time;
import io.bloco.faker.components.University;
import io.bloco.faker.helpers.MapHelper;

public class Faker {

  public static final String DEFAULT_LOCALE = "en";

  public final Address address;
  public final App app;
  public final Avatar avatar;
  public final Book book;
  public final Bool bool;
  public final Business business;
  public final Color color;
  public final Commerce commerce;
  public final Company company;
  public final Date date;
  public final Food food;
  public final Internet internet;
  public final Lorem lorem;
  public final Name name;
  public final Number number;
  public final Placeholdit placeholdit;
  public final PhoneNumber phoneNumber;
  public final SlackEmoji slackEmoji;
  public final Team team;
  public final Time time;
  public final University university;

  private final String locale;
  private final FakerData data;

  public Faker() {
    this(DEFAULT_LOCALE);
  }

  public Faker(String locale) {
    this.locale = locale;

    // Load data
    Map<String, Object> data = loadData(DEFAULT_LOCALE); // Fallbacks first
    if (!this.locale.equals(DEFAULT_LOCALE)) {
      MapHelper.deepMerge(data, loadData(this.locale));
    }
    this.data = new FakerData(data);

    // Load components
    this.address = this.data.getComponent(Address.class);
    this.app = this.data.getComponent(App.class);
    this.avatar = this.data.getComponent(Avatar.class);
    this.book = this.data.getComponent(Book.class);
    this.bool = this.data.getComponent(Bool.class);
    this.business = this.data.getComponent(Business.class);
    this.color = this.data.getComponent(Color.class);
    this.commerce = this.data.getComponent(Commerce.class);
    this.company = this.data.getComponent(Company.class);
    this.date = this.data.getComponent(Date.class);
    this.food = this.data.getComponent(Food.class);
    this.internet = this.data.getComponent(Internet.class);
    this.lorem = this.data.getComponent(Lorem.class);
    this.name = this.data.getComponent(Name.class);
    this.number = this.data.getComponent(Number.class);
    this.placeholdit = this.data.getComponent(Placeholdit.class);
    this.phoneNumber = this.data.getComponent(PhoneNumber.class);
    this.slackEmoji = this.data.getComponent(SlackEmoji.class);
    this.team = this.data.getComponent(Team.class);
    this.time = this.data.getComponent(Time.class);
    this.university = this.data.getComponent(University.class);
  }

  public String getLocale() {
    return locale;
  }

  private Map<String, Object> loadData(String locale) {
    Yaml yaml = new Yaml();
    InputStream input = getDataInputStream(locale);

    Map<String, Object> root = (Map<String, Object>) yaml.load(input);
    Map<String, Object> fakerData = (Map<String, Object>) root.values().iterator().next();
    return (Map<String, Object>) fakerData.get("faker");
  }

  private InputStream getDataInputStream(String locale) {
    InputStream input = getClass().getClassLoader()
        .getResourceAsStream("locales/" + locale + ".yml");

    try {
      if (input != null && input.available() != 0) {
        return input;
      }
    } catch (IOException e) {
    }

    throw new IllegalArgumentException("Unavailable locale \'" + locale + "\'");
  }
}

File: src/main/kotlin/io/bloco/faker/Faker.kt
package io.bloco.faker

import io.bloco.faker.components.Address
import io.bloco.faker.components.App
import io.bloco.faker.components.Artist
import io.bloco.faker.components.Avatar
import io.bloco.faker.components.Book
import io.bloco.faker.components.Bool
import io.bloco.faker.components.Business
import io.bloco.faker.components.Color
import io.bloco.faker.components.Commerce
import io.bloco.faker.components.Company
import io.bloco.faker.components.Date
import io.bloco.faker.components.Food
import io.bloco.faker.components.Internet
import io.bloco.faker.components.Lorem
import io.bloco.faker.components.Name
import io.bloco.faker.components.Number
import io.bloco.faker.components.PhoneNumber
import io.bloco.faker.components.Placeholdit
import io.bloco.faker.components.SlackEmoji
import io.bloco.faker.components.Team
import io.bloco.faker.components.Time
import io.bloco.faker.components.University
import io.bloco.faker.helpers.deepMerge
import org.yaml.snakeyaml.Yaml
import java.io.IOException
import java.io.InputStream

class Faker(val locale: String = DEFAULT_LOCALE) {
    @JvmField
    val address: Address

    @JvmField
    val app: App

    @JvmField
    val artist: Artist

    @JvmField
    val avatar: Avatar

    @JvmField
    val book: Book

    @JvmField
    val bool: Bool

    @JvmField
    val business: Business

    @JvmField
    val color: Color

    @JvmField
    val commerce: Commerce

    @JvmField
    val company: Company

    @JvmField
    val date: Date

    @JvmField
    val food: Food

    @JvmField
    val internet: Internet

    @JvmField
    val lorem: Lorem

    @JvmField
    val name: Name

    @JvmField
    val number: Number

    @JvmField
    val placeholdit: Placeholdit

    @JvmField
    val phoneNumber: PhoneNumber

    @JvmField
    val slackEmoji: SlackEmoji

    @JvmField
    val team: Team

    @JvmField
    val time: Time

    @JvmField
    val university: University
    private val data: FakerData

    init {

        // Load data
        val rawData = loadData(DEFAULT_LOCALE) // Fallbacks first
        if (locale != DEFAULT_LOCALE) {
            (rawData as MutableMap<String, Any>).deepMerge(loadData(locale))
        }
        this.data = FakerData(rawData)

        // Load components
        address = data.getComponent(Address::class)
        app = data.getComponent(App::class)
        artist = data.getComponent(Artist::class)
        avatar = data.getComponent(Avatar::class)
        book = data.getComponent(Book::class)
        bool = data.getComponent(Bool::class)
        business = data.getComponent(Business::class)
        color = data.getComponent(Color::class)
        commerce = data.getComponent(Commerce::class)
        company = data.getComponent(Company::class)
        date = data.getComponent(Date::class)
        food = data.getComponent(Food::class)
        internet = data.getComponent(Internet::class)
        lorem = data.getComponent(Lorem::class)
        name = data.getComponent(Name::class)
        number = data.getComponent(Number::class)
        placeholdit = data.getComponent(Placeholdit::class)
        phoneNumber = data.getComponent(PhoneNumber::class)
        slackEmoji = data.getComponent(SlackEmoji::class)
        team = data.getComponent(Team::class)
        time = data.getComponent(Time::class)
        university = data.getComponent(University::class)
    }

    @Suppress("UNCHECKED_CAST")
    private fun loadData(locale: String): Map<String, Any> {
        val yaml = Yaml()
        val input = getDataInputStream(locale)
        val root = yaml.load(input) as Map<String, Any>
        val fakerData = root.values.iterator().next() as Map<String, Any>
        return fakerData["faker"] as Map<String, Any>
    }

    private fun getDataInputStream(locale: String): InputStream {
        val input = javaClass.classLoader
            .getResourceAsStream("locales/$locale.yml")
        try {
            if (input != null && input.available() != 0) {
                return input
            }
        } catch (e: IOException) {
            throw IllegalArgumentException("Unavailable locale '$locale'", e)
        }
        throw IllegalArgumentException("Unavailable locale '$locale'")
    }

    companion object {
        const val DEFAULT_LOCALE = "en"
    }
}
